### ISSUE TITLE
Tight CPU loop for AUX input is a problem.

### ISSUE THREAD

**Issue by skx** [skx/cpmulator#53]
Running zork1, etc, is fine, because that uses the nice I/O syscall, but running mbasic involves the use of the A_READ/A_WRITE functions and the reader in particular ends up busy-waiting on pending-input, which causes high CPU load (on the host).

We should look at using termbox, or similar, so that we can poll for pending input more efficiently - or alternatively we should poll STDIN in a goroutine and route input to the `cpm` object.

**Comment by skx** [comment]
OK I'm bumping the priority of this issue now, for only one reason:

* A friend tried to deploy this application on a Mac and the console handling was appalling.
  * Keys echoed repeatedly, and high CPU.

Not sure what the solution is, it seems there are two routes to go down:

* Use a goroutine for STDIN reading and a channel for signalling availability.
* Use a curses-like approach where we don't have to change too much - I certainly don't want to keep track of console coordinates, and want string-output to "just work" - but we get better I/O handling for testing whether keyboard input is pending.

**Comment by skx** [comment]
I'll not have time to experiment with this over the next few days, but I've been thinking about what is required.

Looking over my friend's shoulder it seems there are two issues at play here:

* Duplicate characters, which seem to be actually caused by the use of ECHO as input is processed.
* High CPU due to the busy-looping.

The issues seem to be particularly visible when MBASIC.COM is running - so the first step is to confirm that:

1. Run ZORK1
2. Run lihouse
3. Run MBASIC
4. Run TBASIC
5. Run OBASIC

In each case I want to see precisely which syscalls are used, and how they behave.  I think the starting point will be to attach logfiles of each one, being run for a few minutes.  That means I probably need to do two things:

* Log registers on entry to syscalls.
* Avoid the DEBUG environmental variable and instead add a -debug=path/to/file.log CLI argument to allow reproducing things easily.

Once I have the traces I can see which of the calls is the one that are the difficult ones.  I suspect it is the A_READ/A_WRITE as they're the only ones that are used by MBASIC, and were the last to be added.  I can play Zork, complete lihouse, and run the c-compiler (under CCP) happily for hours with no problem here at least.

**Comment by skx** [comment]
So I've had some time to study this and I'm going to commit traces, as per the previous comments.

At the end of the day there's a tradeoff to be made:

* If I want decent non-hot-loops I need to use a wrapper.
   * That makes the output harder, though it makes the input easier.
* I think I can live with the tight CPU.
   *  So the actual side-effects of duplicate characters is the "real" problem here.

And reading CP/M BDOS functions I see what the problem is probably:

* It's all about ECHO vs. NON-ECHO

For example:

* BDOS function 1 (C_READ) - Console input
   * Wait for a character from the keyboard; then echo it to the screen and return it.
* BDOS function 3 - Raw console input
   * Reads a character from the console without checking for ^S / ^Q.
   * i.e. NO ECHO here.
* BDOS function 6 (C_RAWIO) - Direct console I/O
  * Behaviour depends on E
  * E == 0xFF
     * Return a character without echoing if one is waiting; zero if none is available
   * E == 0xFE
      * Return console input status. Zero if no character is waiting, nonzero otherwise.
    * E == 0xFD
       * Wait until a character is ready, return it without echoing.

etc.

So TLDR; I think my echo-handling is not correct.

**Comment by skx** [comment]
Plan!

* Write a test program, in assembly, which will call the various functions and allow interactive testing.
* Once that is ready I will then update the functionality of the syscalls to match expectations - such that it works correctly.

Program will be called "intest[.com]", for "input-test".
